File: J9Ch/src/J_3_Binary_Tree_Divide_Conquer/Optional_11/_246_Binary_Tree_Path_Sum_II_Easy.py
'''/**
 246
 Binary Tree Path Sum II
 '''


"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
"""
class Solution:
    # @param {TreeNode} root the root of binary tree
    # @param {int} target an integer
    # @return {int[][]} all valid paths
    def binaryTreePathSum2(self, root, target):
        # Write your code here
        result = []
        path = []
        if root is None:
            return result
        self.dfs(root, path, result, 0,  target)
        return result

    def dfs(self, root, path, result, l, target):
        if root is None:
            return
        path.append(root.val)
        tmp = target
        for i in xrange(l , -1, -1):
            tmp -= path[i]
            if tmp == 0:
                result.append(path[i:])

        self.dfs(root.left, path, result, l + 1, target)
        self.dfs(root.right, path, result, l + 1, target)

        path.pop()

File: J9Ch/src/J_3_Binary_Tree_Divide_Conquer/Optional_11/_246_Binary_Tree_Path_Sum_II_Easy.java
package J_3_Binary_Tree_Divide_Conquer.Optional_11;
import java.util.*;
/**
 246
 Binary Tree Path Sum II
 * Created by tianhuizhu on 6/28/17.
 */
public class _246_Binary_Tree_Path_Sum_II_Easy {
    class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode(int x) { val = x; }
    }
    public class Solution {
        /**
         * @param root the root of binary tree
         * @param target an integer
         * @return all valid paths
         */
        public List<List<Integer>> binaryTreePathSum2(TreeNode root, int target) {
            // Write your code here
            List<List<Integer>> results = new ArrayList<List<Integer>>();
            ArrayList<Integer> buffer = new ArrayList<Integer>();
            if (root == null)
                return results;
            findSum(root, target, buffer, 0, results);
            return results;
        }

        public void findSum(TreeNode head, int sum, ArrayList<Integer> buffer, int level, List<List<Integer>> results) {
            if (head == null) return;
            int tmp = sum;
            buffer.add(head.val);
            for (int i = level;i >= 0; i--) {
                tmp -= buffer.get(i);
                if (tmp == 0) {
                    List<Integer> temp = new ArrayList<Integer>();
                    for (int j = i; j <= level; ++j)
                        temp.add(buffer.get(j));
                    results.add(temp);
                }
            }
            findSum(head.left, sum, buffer, level + 1, results);
            findSum(head.right, sum, buffer, level + 1, results);
            buffer.remove(buffer.size() - 1);
        }
    }
}
